Functional specs guidance (#1176)
carbon-design-system/carbon-for-ibm-dotcom-website: wiki/Publishing-guidelines/Writing-functional-specs.md

The functional specs document is a deliverable after a functional analysis has been conducted over a full page template or all the way down to proposed component. It offers insights into how a component should be broken down, reused, named, and documents how it should function as a written agreement between design and development. Development will often take these documents and build it based on the requirements, and they will also often be used to help with content strategy from other teams, and also to QA by validate whether something is a bug or not.

<!-- toc start open="true" -->
*{{table of contents auto populates}}*
<!-- toc end -->

### Basic principles
- Reuse existing
- Avoid duplication and redundancy
- Document with consistent
- Design to scale
- Each document and component does one thing, and does that one thing very well
- Don't over engineer
- Start early, listen, and take note

> - Collaborate across disciplines

Author: Maybe we can talk about some of these. I'm not outlining a linear process here just to be clear. I'm outlining guiding principles that effect every aspect of the entire process.

> - Collaborate across disciplines

Reviewer: Sounds good. Grab some time! My calendar is up to date.

### Naming components
There is a document with some basic naming strategy that we try to follow as best we can. Here is the link to our [naming strategy](https://ibm.box.com/s/xlq4romcyp50lnvwuaagyzzbmmesuxen). 

### Data model

> At a low level every single one of our components starts here, and deviates as necessary. Here is a link to the [[Data model]].

Author: It's linked right there at the end. `[[Data model]]`. It's already created. This is merely a reference at the moment. I think more could be said here. Examples given, how to use, and when to deviate. But I'm just trying to give a high level definition. What if we try something like this?

```markdown
### Data model
The [[Data model]] is a shared architecture of content across all of our components, and should be used as a starting point when structuring new components or updating existing components.
```

<br /><br />

## Understanding the template

> The functional spec template is just a set of common tools that I find commonly used across components, and compiled into a single file/template for easy reference and better consistency. All of the sections in this template do not always have to be used, and while I would ask if we could avoid it the fact that we have this template doesn't mean it's the only way or even best way to present a particular component. Use it as a jumping off point, and try your best to stay consistent with it. With all things functional specs related collaboration is important, so we can discuss changing things when reviewing if necessary.

Author: Linking to other wiki files work like this... `[[fs template]]`

> The functional spec template is just a set of common tools that I find commonly used across components, and compiled into a single file/template for easy reference and better consistency. All of the sections in this template do not always have to be used, and while I would ask if we could avoid it the fact that we have this template doesn't mean it's the only way or even best way to present a particular component. Use it as a jumping off point, and try your best to stay consistent with it. With all things functional specs related collaboration is important, so we can discuss changing things when reviewing if necessary.

Reviewer: Cool. Feel free to update!

> The functional spec template is just a set of common tools that I find commonly used across components, and compiled into a single file/template for easy reference and better consistency. All of the sections in this template do not always have to be used, and while I would ask if we could avoid it the fact that we have this template doesn't mean it's the only way or even best way to present a particular component. Use it as a jumping off point, and try your best to stay consistent with it. With all things functional specs related collaboration is important, so we can discuss changing things when reviewing if necessary.

Author: oh look at that, I can update your comment. That's nice...

> The functional spec template is just a set of common tools that I find commonly used across components, and compiled into a single file/template for easy reference and better consistency. All of the sections in this template do not always have to be used, and while I would ask if we could avoid it the fact that we have this template doesn't mean it's the only way or even best way to present a particular component. Use it as a jumping off point, and try your best to stay consistent with it. With all things functional specs related collaboration is important, so we can discuss changing things when reviewing if necessary.

Author: let me actually rename that file. fs-template isn't clear, and it'd be nice to spell it out like you have here.

> The functional spec template is just a set of common tools that I find commonly used across components, and compiled into a single file/template for easy reference and better consistency. All of the sections in this template do not always have to be used, and while I would ask if we could avoid it the fact that we have this template doesn't mean it's the only way or even best way to present a particular component. Use it as a jumping off point, and try your best to stay consistent with it. With all things functional specs related collaboration is important, so we can discuss changing things when reviewing if necessary.

Reviewer: Great idea! It'll save some words! :)

> ![image](https://user-images.githubusercontent.com/3793636/117873919-f6faba80-b265-11eb-81a5-039bdcd822e8.png)  here is a link to the functional specs template: [[fs template]]

<br />

### Meta data

The meta data is used for organizational purposes to help the reader find what they are looking for while also providing the team with language and breakdown of how a component is formed. Some of these items are auto generated, so there is no effort on your part to update while others are manual. 

#### Table of contents

> The table of contents in the template is auto generated. As long as you are using proper markdown headings it should find them just fine, and list the page outline up at the top of the page. So make sure to include the table of contents section template in your component wiki page.

Author: Not necessarily at the top of the page, but more of where you placed the template section

> With correct markdown headings, the page outline will automatically appear within the table of contents section you added.

> The table of contents in the template is auto generated. As long as you are using proper markdown headings it should find them just fine, and list the page outline up at the top of the page. So make sure to include the table of contents section template in your component wiki page.

Reviewer: Just edited the suggestion. Does that work?

#### Dependencies
The dependency list is one of the most important aspects of the functional specs. It communicates to the developer, and creates almost this written agreement between design and development what components should be used when building a component out. 

###### How to add a dependency

1. Find the dependency list in your component's template.
2. Add a new list item
3. Link to the component using the Wiki linking strategy.

<details>
  <summary><strong>Example</strong></summary><br />

```html
- [[Component or wiki page name 1]]
- [[Component or wiki page name 2]]
- [[Component or wiki page name 3]]
```

> ![image](https://user-images.githubusercontent.com/3793636/117873919-f6faba80-b265-11eb-81a5-039bdcd822e8.png)  Notice the double brackets with the file name `[[File name]]` in-between. This is different from a markdown link `[text](url)`.

<br />
</details><br /><br />

#### Used by

> By properly linking dependencies we will also run an automation script to collect all the dependencies and create a used by list in addition. Used by components once populated communicates to the reader where this component is used.

Reviewer: @photodow I'm a little unsure what we are telling the reader to do here. Is there something they need to do in the template? When I first read this, I thought the list was autogenerated based on the automation script you mention but then I read again and wasn't sure. Thanks.

Author: It's similar to the table of contents. There is a section in the template they need to make sure is included if they want to have a used by section. After that it just a matter of making sure the component is properly listed as a dependency in other components.

Author: Not perfect, but let me know if this makes anymore sense...
> Used by components communicates to the reader where this component is used in other components. As long as you include this, and properly link to it in other component's dependencies then this list will create itself.

###### How to add a used by

> ![image](https://user-images.githubusercontent.com/3793636/117873919-f6faba80-b265-11eb-81a5-039bdcd822e8.png)  Keep dependencies up to date, and leverage used by template section in the component wiki page.

#### Backlinks

> Every time a wiki document references a component other than itself it creates a back link. Those back links are listed out on that component's page as a snippet and categorized by pages. When clicked it will go to the section of the page where it was talked about, and highlight its contents. A backlink doesn't have to be used as a depency to work, and this list is automated. As long as you include it on the page it'll render its content if available.

Reviewer: ```suggestion
When a wiki document references another component, it creates a back link. Back links are autogenerated and are listed by page, along with a snippet of text from the page for context. The links take you directly to the mention on the page. 
```

Reviewer: @photodow I wasn't sure what you meant by the dependency comment so I left it out but we can add back in.

Author: I can go either way. I was just trying to say that it ignores the links in the dependency list, and trying to differentiating it from some of the other sections. I think what you have here is fine though.

#### Resources

> The resource section is a numbered list of additional resources not included in the dependency list. For example design specs, box folders, prototypes, and maybe additional documentation are good things you could include here. Follow the numbering standard if possible to star consistent. `r` for resources, and `#` in the list. If for whatever reason we need to remove an item it's best to keep the numbers the name, and skip since there is possible cross page linking and to avoid multiple updates when a resource becomes outdated.

Reviewer: ```suggestion
The resource section is a numbered list of additional resources not included in the dependency list. For example design specs, box folders, prototypes, and any additional documentation that will be helpful. Follow the numbering standard to stay consistent (`r` for resources, and `#` in the list). If you need to remove an item, it's best to keep the numbers the same as there is possible cross page linking. This also avoids multiple updates when a resource becomes outdated.
```

Author: Could we write this sentence more like the following? I think it's important to communicate that by "remaining the same" means that the resource itself doesn't change its reference number, and even if we end up skipping a couple numbers it's ok.

> If you need to remove an item, it's best to keep its number the same. Skipping numbers in this scenario is ok, and usually preferred as there is possible cross page linking.

Reviewer: Yes, this is good. Two slight tweaks.... for `okay` (always full word) and removed the words `usually` and `possible`. The `could be` handles that it might not be the case.   

> If you need to remove an item, it's best to keep its number the same. Skipping numbers in this scenario is okay, and preferred as there could be cross page linking.

<details>
  <summary><strong>Example</strong></summary><br />

```html
- r1: [resource example](www.example.com)
- r2: [resource example](www.example.com)
- r3: [resource example](www.example.com)
- r6: [resource example](www.example.com)
- r7: [resource example](www.example.com)
```

<br />
</details><br /><br />

### Sections

> The sections on the page are just some ways I've used in the past to break content down, and provide the functional specs required for development. Ultimately the following levels use at your discretion, and discuss in review, but the following is how I tend to like to use them.

Reviewer: ```suggestion
The following sections provide a structure for the information required for component development. Using this same structure provides consistency across documents and ease of use for all stakeholders. If there are adjustments needed for your component, we can review and discuss the best approach. 
```

<br /><br />

#### Global

> The global level is often where I'll put a visual of the component, and a high level break down of its elements and components. Every once in while I'll add a global option table or a logic conditional table to demonstrate how whole entire component's functionality can be modified. I typically won't include a markdown title to help indicate that it's apart of the highest level of the component.

Reviewer: ```suggestion
Use Global to include a visual of the component, and a high level break down of its elements and components. Depending on the needs of the component, it may be helpful to include a global option table or a logic conditional table to demonstrate how the whole component's functionality can be modified. 
```

Reviewer: @photodow I didn't understand the last sentence here. 

> I typically won't include a markdown title to help indicate that it's apart of the highest level of the component.

Author: Just trying to communicate that if this section is included there isn't typically a `### global` heading.

It's usually something like this at a high level...

```
# component name

{{meta}
{{global functions}}

## Level 1

{{level 1 functions}}
```

#### Level 1

> Level 1 tends to be the layer of elements, and usually any dependencies if there are any. Here I can identify what high level elements are `optional`, and what `type` (e.g. `string`, `component`) it might be.

Reviewer: ```suggestion
Level 1 usually covers the layer of elements, along with any dependencies. This is where you identify what high level elements are `optional`, and what `type` (e.g. `string`, `component`) it might be.
```

Author: I think I meant to add the word "highest" here? And maybe "outline" makes more sense than covers? 

> Level 1 typically outlines the highest layer of elements within a component, along with any dependencies. This is where you identify what high level elements are `optional`, and what `type` (e.g. `string`, `component`) it might be. Underneath here we can continue to break it down further, or add functional details as needed.

Reviewer: Yes this is great, James. For this sentence: 

> Underneath here we can continue to break it down further, or add functional details as needed.

Do you mean in this section? If so, does this work?

> Use this section to break the content down further, or add functional details, as needed.

Author: I might would suggest we use the word component instead of content because it can be a little more than that, but yeah that works for me.
